bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go

```diff
@@ -95,6 +98,31 @@ type KubeadmConfigSpec struct {
 	// For more information, refer to https://github.com/kubernetes-sigs/cluster-api/pull/2763#discussion_r397306055.
 	// +optional
 	UseExperimentalRetryJoin bool `json:"useExperimentalRetryJoin,omitempty"`
+
+	// IgnitionConfig contains Ignition specific configuration.
+	//
+	// If set, bootstrap data will be generated in Ignition format instead of cloud-init.
+	//
+	// +optional
+	IgnitionConfig *IgnitionConfig `json:"ignitionConfig,omitempty"`
+}
+
+// IgnitionConfig contains Ignition specific configuration.
+type IgnitionConfig struct {
+	// ContainerLinuxConfig contains CLC specific configuration.
+	//
+	// +optional
+	ContainerLinuxConfig *ContainerLinuxConfig `json:"containerLinuxConfig,omitempty"`
+}
+
+// ContainerLinuxConfig contains CLC specific configuration.
+//
+// We use structured type here to keep space for adding additional fields, for example 'version'.
+type ContainerLinuxConfig struct {
+	// AdditionalConfig contains additional configuration which will be added bootstrap data in CLC format.
+	//
+	// +optional
+	AdditionalConfig string `json:"additionalConfig"`
```

Comment:
Do we want to use an unstructured string here?

Reply:
From https://github.com/kubernetes-sigs/cluster-api/issues/3430#issuecomment-776643636:
>additionalConfig field will be of type string to avoid pulling entire Ignition config scheme into KubeadmControlPlane CRD and also to allow using different transpiler versions in the future if desired.

Reply:
I think this code makes a reasonable choice by using a higher-level front-end language here, rather than embedding an Ignition config directly.  Note, though, that the major reason we never did this sort of thing with CLCs is that CLCs are unversioned.  It's impossible to tell by inspection what version of the Ignition spec a CLC will be transpiled into (and thus what OS versions it'll run on), or what CLC features are supported.  That's entirely a property of the transpiler version.

Reply:
@vincepri can we mark this as resolved?